posix/unistd.c

```diff
@@ -0,0 +1,609 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/corelibc/errno.h>
+#include <openenclave/corelibc/limits.h>
+#include <openenclave/corelibc/pthread.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/string.h>
+#include <openenclave/corelibc/sys/stat.h>
+#include <openenclave/corelibc/sys/utsname.h>
+#include <openenclave/corelibc/unistd.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/time.h>
+#include <openenclave/internal/trace.h>
+#include "mount.h"
+#include "posix_t.h"
+
+int oe_gethostname(char* name, size_t len)
+{
+    int ret = -1;
+    struct oe_utsname uts;
+
+    if ((ret = oe_uname(&uts)) != 0)
+        OE_RAISE_ERRNO(oe_errno);
+
+    oe_strlcpy(name, uts.nodename, len);
+    ret = 0;
+
+done:
+    return ret;
+}
+
+int oe_getdomainname(char* name, size_t len)
+{
+    int ret = -1;
+    struct oe_utsname uts;
+
+    if ((ret = oe_uname(&uts)) != 0)
+        OE_RAISE_ERRNO(oe_errno);
+
+    oe_strlcpy(name, uts.domainname, len);
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+static char _cwd[OE_PATH_MAX] = "/";
+static oe_pthread_spinlock_t _lock;
+
+char* oe_getcwd(char* buf, size_t size)
+{
+    char* ret = NULL;
+    char* p = NULL;
+    size_t n;
+    bool locked = false;
+
+    if (buf && size == 0)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (!buf)
+    {
+        n = OE_PATH_MAX;
+        p = oe_malloc(n);
+
+        if (!p)
+            OE_RAISE_ERRNO(OE_ENOMEM);
+    }
+    else
+    {
+        n = size;
+        p = buf;
+    }
+
+    oe_pthread_spin_lock(&_lock);
+    locked = true;
+
+    if (oe_strlcpy(p, _cwd, n) >= n)
+        OE_RAISE_ERRNO(OE_ERANGE);
+
+    oe_pthread_spin_unlock(&_lock);
+    locked = false;
+
+    ret = p;
+    p = NULL;
+
+done:
+
+    if (locked)
+        oe_pthread_spin_unlock(&_lock);
+
+    if (p && p != buf)
+        oe_free(p);
+
+    return ret;
+}
+
+int oe_chdir(const char* path)
+{
+    int ret = -1;
+    char real_path[OE_PATH_MAX];
+    struct oe_stat st;
+    bool locked = false;
+
+    /* Resolve to an absolute canonical path. */
+    if (!oe_realpath(path, real_path))
+        return -1;
+
+    /* Fail if unable to stat the path. */
+    if (oe_stat(real_path, &st) != 0)
+    {
+        // oe_errno set by oe_stat().
+        return -1;
+    }
+
+    /* Fail if path not a directory. */
+    if (!OE_S_ISDIR(st.st_mode))
+    {
+        oe_errno = OE_ENOTDIR;
+        return -1;
+    }
+
+    /* Set the _cwd global. */
+    oe_pthread_spin_lock(&_lock);
+    locked = true;
+
+    if (oe_strlcpy(_cwd, real_path, OE_PATH_MAX) >= OE_PATH_MAX)
+        OE_RAISE_ERRNO(OE_ENAMETOOLONG);
+
+    oe_pthread_spin_unlock(&_lock);
```

Comment:
Do we need this here? we cold go to done and release the lock.

Reply:
Nice simplification. Thanks. Fixed.